SQLRepeatRecord

kaapstorm opened the pull request:
## Summary

These changes are guided by [[CEP] Migrate RepeaterRecord to SQL](https://github.com/dimagi/commcare-hq/issues/28314).

This PR establishes the groundwork for using SQL for data forwarding. It only adds models; it does not change any behavior.

(I am creating this PR as a draft because its base branch is PR #28948. When that PR is merged, I'll switch this to a real PR.)

cc @calellowitz @orangejenny @millerdev @snopoke 


## Safety Assurance

- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I am certain that this PR will not introduce a regression for the reasons below

### Automated test coverage
Tests included in PR


### QA Plan
Not yet applicable


### Safety story
This PR does not change any behavior.


### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

kaapstorm committed:
SQLRepeatRecord and related models

kaapstorm committed:
⚖ RepeaterManagerTests

kaapstorm committed:
⚖ Test repeat record ordering

kaapstorm committed:
RepeaterManager

millerdev commented on corehq/motech/repeaters/migrations/0002_sqlrepeatrecord.py:
```diff
@@ -0,0 +1,113 @@
+                ('id', models.AutoField(auto_created=True,
+                                        primary_key=True,
+                                        serialize=False,
+                                        verbose_name='ID')),
```
This is much prettier formatting that the default generated by the `makemigrations` command, but something is sticking in my memory that Django will want to reformat this file each time `makemigrations` is run on this app. Is that true, or am I mis-remembering?  (I'd be happy I'm wrong)

orangejenny replied:
This should be okay - running `makemigrations` on future changes won't affect the formatting of existing migrations.

kaapstorm replied:
Good question! I tested. When Django creates a new migration, it does not touch existing migrations.

millerdev reviewed (commented):
Looks good to me. :heart: 1970-01-06 comment

orangejenny commented:
So exciting to see this happening 🥇 

I didn't read the code closely but had a comment on naming: are you planning on keeping the `SQL` prefixes for class names? I've gotten critical feedback on that when doing migrations (and I've come around to that point of view, now I see the `SQL` as cruft). It's pretty painless to make the prefix temporary if you set `db_table` to match the un-prefixed name, then later you can remove the prefix, remove the old couch class, and remove the `db_table` in a cleanup PR.

kaapstorm commented:
> are you planning on keeping the `SQL` prefixes for class names? I've gotten critical feedback on that when doing migrations

Good call!

Do you think `RepeaterStub` is a good name for `SQLRepeaterStub`? I'm not delighted by either, but I couldn't think of anything better.

kaapstorm committed:
Set `db_table`

kaapstorm committed:
Split multicolumn index

based on [PostgreSQL](https://www.postgresql.org/docs/9.6/indexes-multicolumn.html) [docs](https://www.postgresql.org/docs/9.6/indexes-bitmap-scans.html)

kaapstorm committed:
Misleading

kaapstorm committed:
Update migration

orangejenny commented:
I think `RepeaterStub` is okay. The only other thing that comes to mind is `RepeaterJoiner` but I don't think that's obviously better.

kaapstorm committed:
Rename SQLRepeaterStub to RepeaterLink

kaapstorm commented:
Urgh. I renamed it to RepeaterLink. I'm not sure it's an improvement, but I am sure I'm overthinking this. :joy:

orangejenny commented:
Let me join you in overthinking...I don't really love `RepeaterLink` because "link" already has a meaning in web applications. I'd like to rename linked apps/domains for this same reason, we have some fairly confusing UI code where we're making `<a>` links that related to linked domains. Not sure if that will come up with repeaters, but don't repeaters have URL attributes (which, granted, we probably call urls and not links)?

kaapstorm commented:
That is a good point. OK, I'll go with `RepeaterStub`.

kaapstorm committed:
Rename RepeaterLink to RepeaterStub

kaapstorm commented:
(I kinda prefer `RepeaterJoiner` to `RepeaterStub`, but I just couldn't do `RepeaterJoinerManager`. I even (briefly) considered `Peter` (very, very briefly).)

proteusvacuum commented on corehq/motech/repeaters/models.py:
```diff
@@ -164,6 +167,44 @@ def log_repeater_success_in_datadog(domain, status_code, repeater_type):
+        """
+        return (
+            self.get_queryset()
+            # Is not paused
```
nit: I think this comment is redundant with the code. I think you wanted to make this code more readable, you could add in some named variables instead

kaapstorm replied:
Ooo! I like that idea! 2d8607d

proteusvacuum reviewed (approved):
🎉

kaapstorm committed:
♻ Replace comments with variables

kaapstorm committed:
Merge branch 'master' into nh/rep/sqlrepeatrecord

kaapstorm committed:
Domain deletion

kaapstorm committed:
Include in domain dump

Should we really be dumping RepeatRecord and SQLRepeatRecord?

kaapstorm committed:
Update index hash

kaapstorm commented on corehq/apps/dump_reload/sql/dump.py:
```diff
@@ -139,6 +139,12 @@
     FilteredModelIteratorBuilder('case_importer.CaseUploadRecord', SimpleFilter('domain')),
+    FilteredModelIteratorBuilder('motech.ConnectionSettings', SimpleFilter('domain')),
+    FilteredModelIteratorBuilder('repeaters.RepeaterStub', SimpleFilter('domain')),
+    # NH (2021-01-08): Including SQLRepeatRecord because we dump (Couch)
+    # RepeatRecord, but this does not seem like a good idea.
+    FilteredModelIteratorBuilder('repeaters.SQLRepeatRecord', SimpleFilter('domain')),
+    FilteredModelIteratorBuilder('repeaters.SQLRepeatRecordAttempt', SimpleFilter('repeat_record__domain')),
```
For 99% of domains, including repeat records with dumped domain data is insignificant, but for a handful it could be gigantic, with very little benefit. I suspect that in the future, (SQL) repeat records will be deleted automatically after six to twelve months anyway.

I'd love to hear opinions on this, especially if you have dealt with domain migrations. (@esoergel @snopoke ?)

snopoke replied:
We can certainly exclude all `SQLRepeatRecordAttempt` objects. Is it possible to only include `SQLRepeatRecord` objects that have not been successful yet?

kaapstorm replied:
Good idea. Yeah, we can filter by `state`.

kaapstorm commented:
ping @dannyroberts for context
